File: micros/triangles.cc
/**
 * Counting, and optionally listing, triangles in a graph
 *
 * This application counts, and optionally lists, the number of
 * triangles in a given graph.  When the counting has completed, it
 * prints the number of triangles found.  If the '-v' command line
 * argument is present, the application also lists the triangles found
 * (one triangle per line, tab-separated identifiers).
 *
 * Based on an algorithm described here
 *
 *   Thomas Schank and Dorothea Wagner.  Finding, Counting and Listing
 *   All Triangles in Large Graphs, an Experimental Study.  Technical
 *   report.  University of Karlsruhe, Germany, 2005.
 *   http://i11www.iti.uni-karlsruhe.de/extra/publications/sw-fclt-05_t.pdf
 *
 * and here
 *
 *   Thomas Schank and Dorothea Wagner.  Finding, Counting and Listing
 *   All Triangles in Large Graphs, an Experimental Study.  In
 *   Proceedings of the Fourth International Conference on
 *   Experimental and Efficient Algorithms, pp. 606-609, Santorini
 *   Island, Greece, 2005.
 */

#include <sys/time.h>
#include <sys/resource.h>
#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <string.h>
#include "pmgd.h"
#include "util.h"
#include "neighbor.h"
#include "time.h"

using namespace std;
using namespace PMGD;

/**
 * Program name
 */
char *prog_name = NULL;

/*
 * Graph property names
 */
static const char ID[] = "pmgd.loader.id";

/**
 * Set of node IDs
 *
 * Set expected to be small on average.
 */
typedef set<Node *> id_set_t;

/**
 * Auxiliary structure used for sorting nodes by their degrees
 */
class DegreeMap {
public:
    struct Degree {
        Node *node;
        long long id;  // For compatibility with Java version
        long long degree;
        id_set_t id_set;

        /**
         * Sort by the degree first and then by id
         * @param d instance of Degree to compare against this
         * @return true if this is "smaller"
         * Higher degree comes first and if there is a tie, the lower id
         * wins.
         */
        bool operator <(const Degree &d) const
        {
            return degree > d.degree
                || (degree == d.degree && id < d.id);
        }
    };

private:
    struct set_sort {
        bool operator()(Degree *const &d1, Degree *const &d2)
            { return *d1 < *d2; }
    };

    typedef std::map<const long long, Degree> M;
    M _node_map;
    std::set<Degree *, set_sort> _degree_set;

public:
    void add(Node *n, long long id, long long d)
    {
        M::iterator mi = _node_map.insert(M::value_type(id, Degree{n, id, d})).first;
        _degree_set.insert(&mi->second);
    }

    std::set<Degree *>::const_iterator begin() const
    {
        return _degree_set.begin();
    }

    std::set<Degree *>::const_iterator end() const
    {
        return _degree_set.end();
    }

    Degree &operator[](long long id) { return _node_map[id]; }
};

/**
 * Print information about how to invoke this program
 * @param o the put-to stream
 */
void print_usage(ostream &o)
{
    o << "Usage: " << prog_name << " [OPTION]... GRAPH\n";
    o << "\n";
    o << "  -h  print this help and exit\n";
    o << "  -v  be verbose and output the triangles\n";
    o << "  -t  print elapsed time and usage\n";
}

/**
 * The application's main function
 * @param argc the number of command line arguments
 * @param argv the array of command line arguments
 * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
 */
int main(int argc, char *argv[])
{
    bool verbose = false;
    long long num_nodes = 0;
    long long num_triangles = 0;

    prog_name = argv[0];
    char *graph_name;

    bool timing = false;
    struct timeval t1, t2;
    struct rusage u1, u2;

    // Parse the command line arguments
    int argi = 1;
    while (argi < argc && argv[argi][0] == '-') {
        if (strcmp(argv[argi], "-h") == 0) {
            print_usage(cout);
            return 0;
        }
        else if (strcmp(argv[argi], "-v") == 0) {
            verbose = true;
        }
        else if (strcmp(argv[argi], "-t") == 0) {
            timing = true;
        }
        else {
            cerr << prog_name << ": " << argv[argi] << ": Unrecognized option\n";
            return 1;
        }
        ++argi;
    }
    if (argi + 1 > argc) {
        cerr << prog_name << ": No input graph\n";
        print_usage(cerr);
        return 1;
    }
    graph_name = argv[argi];

    try {
        // Open the graph.
        Graph db(graph_name, Graph::ReadOnly);

        // Start a transaction.
        Transaction tx(db);

        // StringID for node id.
        StringID id_sid("pmgd.loader.id");

        // Time and usage measurements.
        (void)getrusage(RUSAGE_SELF, &u1);
        (void)gettimeofday(&t1, NULL);

        // The degree of each node.
        DegreeMap degrees;
        for (NodeIterator n = db.get_nodes(); n; n.next()) {
            ++num_nodes;
            long long degree = 0;
            for (EdgeIterator e = n->get_edges(); e; e.next())
                ++degree;
            degrees.add(&*n, db.get_id(*n), degree);
        }

        // Fill a vector with id and degree information and sort it.
        //
        // While perhaps not apparent upon first reading the algorithm
        // described in Schank and Wagner's paper, it appears that the
        // nodes in the graph must not be compared according their IDs
        // but rather by their positions in the sorted list.

        // Find the triangles.
        for (const auto &s : degrees) {
            for (NodeIterator ti = get_neighbors(*s->node); ti; ti.next()) {
                auto &t = degrees[db.get_id(*ti)];
                if (*s < t) {
                    for (auto n : s->id_set) {
                        if (t.id_set.find(n) != t.id_set.end()) {
                            ++num_triangles;
                            if (verbose) {
                                cout << n->get_property(id_sid).int_value()
                                    << "\t"
                                    << s->node->get_property(id_sid).int_value()
                                    << "\t"
                                    << t.node->get_property(id_sid).int_value()
                                    << "\n";
                            }
                        }
                    }
                    t.id_set.insert(s->node);
                }
            }
        }

        // Time and usage measurements.
        (void)gettimeofday(&t2, NULL);
        (void)getrusage(RUSAGE_SELF, &u2);
    }
    catch (Exception e) {
        print_exception(e, stderr);
        return 2;
    }

    cout << num_triangles << "\n";

    if (timing) print_delta(cout, t1, t2, u1, u2);

    return 0;
}

File: benchmarks/microbenchmarks/Triangles.java
/**
 * Counting, and optionally listing, triangles in a graph
 *
 * This application counts, and optionally lists, the number of
 * triangles in a given graph.  When the counting has completed, it
 * prints the number of triangles found.  If the '-v' command line
 * argument is present, the application also lists the triangles found
 * (one triangle per line, tab-separated identifiers).
 *
 * Based on an algorithm described here
 *
 *   Thomas Schank and Dorothea Wagner.  Finding, Counting and Listing
 *   All Triangles in Large Graphs, an Experimental Study.  Technical
 *   report.  University of Karlsruhe, Germany, 2005.
 *   http://i11www.iti.uni-karlsruhe.de/extra/publications/sw-fclt-05_t.pdf
 *
 * and here
 *
 *   Thomas Schank and Dorothea Wagner.  Finding, Counting and Listing
 *   All Triangles in Large Graphs, an Experimental Study.  In
 *   Proceedings of the Fourth International Conference on
 *   Experimental and Efficient Algorithms, pp. 606-609, Santorini
 *   Island, Greece, 2005.
 */

import java.util.*;
import java.lang.String;
import java.lang.Comparable;
import java.lang.management.*;
import pmgd.*;

public class Triangles {
    /**
     * Program name
     */
    String progName = "Triangles";

    /**
    * Auxiliary structure used for sorting nodes by their degrees
    */
    public class Degree implements Comparable<Degree>
    {
        Node _n;
        long _id;      ///< Node ID
        long _degree;   ///< Node's number of incident edges
        public Set<Long> id_set;

        public Degree(Node n, long id, long deg)
        {
            _n = n;
            _id = id;
            _degree = deg;
            id_set = new HashSet<Long>();
        }

        public Node getNode()
        {
            return _n;
        }

        public long getDegree()
        {
            return _degree;
        }

        public long getId()
        {
            return _id;
        }

        /**
        * Sort by the degree first and then by id
        * @param other instance of Degree to compare against this
        * @return true if this is "smaller"
        * Higher degree comes first and if there is a tie, the lower id
        * wins.
        */
        public int compareTo(Degree other)
        {
            int retVal;
            if (_degree == other.getDegree())
                retVal = (int)(_id - other.getId());
            else
                retVal = (int)(other.getDegree() - _degree);
            return retVal;
        }
    }

    public class DegreeMap
    {
        Map<Long,Degree> _node_map;
        Set<Degree> _degree_set;

        public class DegreeCompare implements Comparator<Degree>
        {
            @Override
            public int compare(Degree d1, Degree d2)
            {
                return d1.compareTo(d2);
            }
        }

        public DegreeMap()
        {
            _node_map = new HashMap<Long,Degree>();
            _degree_set = new TreeSet<Degree>(new DegreeCompare());
        }

        public void add(Node n, long id, long d)
        {
            Degree deg = new Degree(n, id, d);
            _node_map.put(id, deg);
            _degree_set.add(deg);
        }

        public Degree get(long id)
        {
            return _node_map.get(id);
        }

        public Set<Degree> get()
        {
            return _degree_set;
        }
    }

    /**
     * The application's main function
     * @param argc the number of command line arguments
     * @param argv the array of command line arguments
     * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
     */
    public void computeTriangles(String[] argv)
    {
        boolean verbose = false;
        int num_nodes = 0;
        long num_triangles = 0;

        String graph_name;

        boolean timing = false;
        int argc = argv.length;

        // Parse the command line arguments
        int argi = 0;
        while (argi < argc && argv[argi].charAt(0) == '-') {
            if (argv[argi].compareTo("-h") == 0) {
                printUsage();
                return;
            }
            else if (argv[argi].compareTo("-t") == 0) {
                timing = true;
            }
            else if (argv[argi].compareTo("-v") == 0) {
                verbose = true;
            }
            else {
                System.out.println(progName + ": " + argv[argi] + ": Unrecognized option");
                return;
            }
            argi += 1;
        }
        if (argi + 1 > argc) {
            System.out.println(progName + ": No input graph");
            printUsage();
            return;
        }
        graph_name = argv[argi];
        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        long userTime, cpuTime;

        try {
            // Open the graph.
            Graph db = new Graph(graph_name, Graph.OpenOptions.ReadOnly);

            // Start a transaction.
            Transaction tx = new Transaction(db, false, true);
            StringID ID = new StringID("pmgd.loader.id");

            // Time measurement.
            long startTime = bean.getCurrentThreadUserTime();
            long startTotal = bean.getCurrentThreadCpuTime();

            DegreeMap degrees = new DegreeMap();
            for (NodeIterator n = db.get_nodes(); !n.done(); n.next()) {
                ++num_nodes;
                long degree = 0;
                Node ni = n.get_current();
                long id = db.get_id(ni);
                for (EdgeIterator e = ni.get_edges(); !e.done(); e.next())
                    ++degree;
                degrees.add(ni, id, degree);
            }

            // Fill a vector with id and degree information and sort it.
            //
            // While perhaps not apparent upon first reading the algorithm
            // described in Schank and Wagner's paper, it appears that the
            // nodes in the graph must not be compared according their IDs
            // but rather by their positions in the sorted list.
            // Find the triangles.
            for (Degree s:degrees.get()) {
                Node sin = s.getNode();
                long id_sin = db.get_id(sin);
                for (NodeIterator ti = sin.get_neighbors(); !ti.done(); ti.next()) {
                    Node tin = ti.get_current();
                    long id_tin = db.get_id(tin);
                    Degree t = degrees.get(id_tin);
                    if (s.compareTo(t) < 0) {
                        for (long nid : s.id_set) {
                            if (t.id_set.contains(nid)) {
                                ++num_triangles;
                                if (verbose) {
                                    System.out.println(degrees.get(nid).getNode().get_property(ID).int_value() + "\t" +
                                                    sin.get_property(ID).int_value() + "\t" +
                                                    tin.get_property(ID).int_value());
                                }
                            }
                        }
                        t.id_set.add(s.getId());
                    }
                }
            }

            // Time measurement.
            long endTime = bean.getCurrentThreadUserTime();
            long endTotal = bean.getCurrentThreadCpuTime();
            userTime = (endTime - startTime)/1000;
            cpuTime = (endTotal - startTotal)/1000;
        }
        catch (pmgd.Exception e) {
            e.print();
            return;
        }

        System.out.println(num_triangles);

        if (timing) {
            long sysTime = cpuTime - userTime;
            float upercent = 100 * (userTime / (float)cpuTime);
            float spercent = 100 * (sysTime / (float)cpuTime);
            System.out.println("Elapsed time is " + cpuTime + " microseconds");
            System.out.println("User time is " + userTime + " microseconds (" + upercent + "%)");
            System.out.println("System time is " + sysTime + " microseconds (" + spercent + "%)");
        }

        return;
    }

    // Trouble using the Degree class due to static nature of
    // main.
    public static void main(String[] argv)
    {
        Triangles t = new Triangles();
        t.computeTriangles(argv);
    }

    /**
     * Print information about how to invoke this program
     * @param o the put-to stream
     */
    void printUsage()
    {
        System.out.println("Usage: " + progName + " [OPTION]... GRAPH");
        System.out.println(" -h  print this help and exit");
        System.out.println(" -v  be verbose and output the triangles");
        System.out.println(" -t  time the execution");
    }
}
